lib/spack/spack/cmd/diff.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+import spack.cmd
+import spack.cmd.common.arguments as arguments
+import spack.environment as ev
+import spack.util.environment
+
+import llnl.util.tty as tty
+from llnl.util.tty.colify import colify
+import spack.util.spack_json as sjson
+import operator
+import sys
+
+description = "compare two specs"
+section = "extensions"
+level = "short"
+
+
+def setup_parser(subparser):
+    """Parser is only constructed so that this prints a nice help
+       message with -h. """
+    arguments.add_common_arguments(
+        subparser, ['specs'])
+
+    subparser.add_argument(
+        '--json',
+        action='store_true',
+        default=False,
+        dest='dump_json',
+        help="Dump json output instead of pretty printing."
+    )
+    subparser.add_argument(
```

Comment:
I'm not sure about this flag, since there are other commands that currently just error when not a single spec matches. Isn't the first result entirely arbitrary?

I'm thinking it would be nice to have a simple CLI menu option when spack is run in an interactive shell so the user can select a specific match. That could be used in `spack uninstall` when `--all` is not provided; as well as selecting one of the > 1 top-level specs from an environment when running `spack -e my_env build-env -- bash`, etc.

Reply:
The parameter `--first` is used to disambiguate the spec - so for example, if I have multiple versions of mpileaks and I ask to diff that against something else, it will raise that disambiguation error and tell me to choose one. If I select first, then it just grabs the first result and gives me the difference, which (as a lazy person sometimes) is what I want. So I don't think it's entirely useless. I verified what I just walked through in testing.

Your second comment is an interesting idea, but it's out of scope for this PR. I'm not doing anything with an interactive spack shell, nor am I touching uninstall or similar. If it's something that could be useful for diff, maybe you could write it up in an issue and i can take a look after this PR?

Reply:
We also need the disambiguate command because it's going to show the user the different options they have in detail (hashes, etc.) which is exactly what they need if they want to diff similar specs.

Update: I added a few more lines to the test file to cover the case of != 2 specs provided.

Reply:
I don't see how `--first` is useful though, since as a user you are probably uninformed about the order in which spack sorts matching specs, and it's not helpful if your spec is not listed first. I would probaby run the command without `--first` by default, and when it fails and outputs a bunch of matches, I would run it again but using the hash from the previous output; that's about as much work as writing `--first` and more flexible?

And about a menu option: it's definitely out of the scope of this PR, but I just commented to point out this multiple-spec-match-error is a recurring issue across different commands and we can make the user experience better.

Reply:
If that argument is valid, then we would need to remove it also from load.py, where the user is also in the same boat. I added it here based on the same logic that it is there.  I just looked at the commit history, and I think the automated use case is a valid one: https://github.com/spack/spack/pull/15622. spack diff could be run in an automated fashion just the same as spack load.

Reply:
Hm yeah, but that use case is about loading a build dependency of an arbitrary version where all would suffice. In this case you want to compare two very specific things, not an arbitrary first match.

Reply:
I also don't feel strongly enough to argue to keep it if the majority think its not useful. Could we ask for others' opinions and get more feedback? If more folks are in agreement with you I'm definitely okay just removing it. I do think there is a use case, but you might be right that we shouldn't develop for this use case until it actually presents itself.